Implement detection of IOMMU (#136)

marquiz opened the pull request:
Add a new feature source, i.e. 'iommu', which detects if an IOMMU is
present and enabled in the kernel. The new node label is
  node.alpha.kubernetes-incubator.io/nfd-iommu-present

Implements #136

marquiz commented:
I verified this, getting "node.alpha.kubernetes-incubator.io/nfd-iommu-enabled=true" label when iommu was enabled in hw and in kernel

kad commented on main.go:
```diff
@@ -136,7 +137,7 @@ func argsParse(argv []string) (args Args) {
   --version                   Output version and exit.
   --sources=<sources>         Comma separated list of feature sources.
-                              [Default: cpuid,rdt,pstate,memory,network,storage,selinux]
+                              [Default: cpuid,iommu,rdt,pstate,memory,network,storage,selinux]
```
this list probably should be sorted alphabetically ?

marquiz replied:
Yeah, I actually did this, but decided to move that into a separate patch. The order here is a leftover from that. There's no PR about that, yet.

balajismaniam replied:
In the end, labels are represented as `map[string]string` and maps are unordered in Go. But I agree that we should sort the sources for readability.

kad commented on main.go:
```diff
@@ -186,6 +187,7 @@ func configureParameters(sourcesWhiteList []string, labelWhiteListStr string) (e
 	// Configure feature sources.
 	allSources := []source.FeatureSource{
 		cpuid.Source{},
+		iommu.Source{},
```
here is practically same question: is the order of the elements matter ? if not, should it be added in the end then ?

marquiz replied:
Same here

marquiz commented:
There's now a separate PR (#139) about arranging the feature sources alphabetically

balajismaniam commented on source/iommu/iommu.go:
```diff
@@ -0,0 +1,46 @@
+
+	// Check if any iommu devices are available
+	devices, err := ioutil.ReadDir("/sys/class/iommu/")
+	if err == nil {
```
The idiomatic way in Go is to check `if err != nil` here. It also removes the dangling else.

balajismaniam commented on source/iommu/iommu.go:
```diff
@@ -0,0 +1,46 @@
+	devices, err := ioutil.ReadDir("/sys/class/iommu/")
+	if err == nil {
+		if len(devices) > 0 {
+			glog.Errorf("Detected %v IOMMU devices", len(devices))
```
This is not an error. Use `Infof` instead?

marquiz replied:
Infof are not printed and thus not shown in logs

balajismaniam commented on source/iommu/iommu.go:
```diff
@@ -0,0 +1,46 @@
+			glog.Errorf("Detected %v IOMMU devices", len(devices))
+			features = append(features, "enabled")
+		} else {
+			glog.Errorf("No IOMMU devices detected")
```
This is not an error. Use Infof instead? May be it is not necessary to log this. It is sufficient if we log when we find a device.

marquiz replied:
As with the previous case, Infof are not shown at all. It is debatable if this should be logged. There are two kinds of feature sources in nfd, some do not log anything if they fail to detect, some print something in any case. I kind of find it useful to see at least something in the log, to verify that the module was run successfully.

marquiz commented:
@balajismaniam see the new version

I hope I addressed your concerns
- I rebased on top of #139 so you see that change, too
- changed the error handling

However, I didn't change the `glog.Errorf()` usage as `glog.Infof` messages are not shown at all. I think changing the logging of the feature sources is worth a separate PR. I think glog is not a good choice for NFD because you cannot change the default log level without bringing in all the logging-related command line flags, which is an overkill considering the simple needs of NFD, IMO.

marquiz commented:
Added explanation to README

balajismaniam commented on source/iommu/iommu.go:
```diff
@@ -0,0 +1,48 @@
+	}
+
+	if len(devices) > 0 {
+		glog.Errorf("Detected %v IOMMU devices", len(devices))
```
Ok. But I think this log is not required. Can we remove this?

balajismaniam commented on source/iommu/iommu.go:
```diff
@@ -0,0 +1,48 @@
+		glog.Errorf("Detected %v IOMMU devices", len(devices))
+		features = append(features, "enabled")
+	} else {
+		glog.Errorf("No IOMMU devices detected")
```
Ok. But I think this log is not required. Can we remove this?

marquiz commented:
> Please rebase and remove the comments. LGTM after that.

@balajismaniam done

marquiz commented:
@balajismaniam: btw, please do not squash the commits, if possible. I find it useful to preserve the commit messages
